File: sdk/ruby/lib/rocksky/resources/scrobble.rb
module Rocksky
  module Resources
    # `app.rocksky.scrobble.*` endpoints.
    class Scrobble < Base
      # Create a new scrobble. Requires an authenticated client.
      #
      # Required: `title`, `artist`. Everything else is optional metadata that
      # gets forwarded as-is to the lexicon's createScrobble body. Extra fields
      # may be passed via `**extra`.
      #
      # Example:
      #
      #   client.scrobble.create_scrobble(
      #     title: "In Bloom",
      #     artist: "Nirvana",
      #     album: "Nevermind",
      #     timestamp: Time.now.to_i
      #   )
      def create_scrobble(title:, artist:, **extra)
        body = { title: title, artist: artist }.merge(extra).compact
        procedure("app.rocksky.scrobble.createScrobble", body: body)
      end

      # Fetch a scrobble by URI.
      def get_scrobble(uri:)
        query("app.rocksky.scrobble.getScrobble", uri: uri)
      end

      # List scrobbles.
      def get_scrobbles(did: nil, following: nil, limit: nil, offset: nil)
        query("app.rocksky.scrobble.getScrobbles",
              did: did, following: following, limit: limit, offset: offset)
      end
    end
  end
end

File: sdk/python/src/rocksky/resources/scrobble.py
"""``app.rocksky.scrobble.*`` — scrobble feed and creation."""

from __future__ import annotations

from typing import Any

from ..models import Scrobble
from ._base import Resource, parse_list, parse_model


class ScrobbleResource(Resource):
    async def get(self, uri: str) -> Scrobble:
        """Get a single scrobble by AT-URI."""
        data = await self._transport.query(
            "app.rocksky.scrobble.getScrobble", params={"uri": uri}
        )
        return parse_model(Scrobble, data)

    async def list(
        self,
        *,
        did: str | None = None,
        following: bool | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> list[Scrobble]:
        """List scrobbles, optionally filtered by actor or restricted to the
        authenticated user's follow graph."""
        data = await self._transport.query(
            "app.rocksky.scrobble.getScrobbles",
            params={
                "did": did,
                "following": following,
                "limit": limit,
                "offset": offset,
            },
            auth=following is not None,
        )
        return parse_list(Scrobble, data, key="scrobbles")

    async def create(
        self,
        *,
        title: str,
        artist: str,
        album: str | None = None,
        duration: int | None = None,
        mb_id: str | None = None,
        isrc: str | None = None,
        album_art: str | None = None,
        track_number: int | None = None,
        release_date: str | None = None,
        year: int | None = None,
        disc_number: int | None = None,
        lyrics: str | None = None,
        composer: str | None = None,
        copyright_message: str | None = None,
        label: str | None = None,
        artist_picture: str | None = None,
        spotify_link: str | None = None,
        lastfm_link: str | None = None,
        tidal_link: str | None = None,
        apple_music_link: str | None = None,
        youtube_link: str | None = None,
        deezer_link: str | None = None,
        timestamp: int | None = None,
    ) -> dict[str, Any]:
        """Create a new scrobble. Requires auth.

        ``timestamp`` is Unix seconds; if omitted the server stamps it.
        """
        body = _drop_none(
            {
                "title": title,
                "artist": artist,
                "album": album,
                "duration": duration,
                "mbId": mb_id,
                "isrc": isrc,
                "albumArt": album_art,
                "trackNumber": track_number,
                "releaseDate": release_date,
                "year": year,
                "discNumber": disc_number,
                "lyrics": lyrics,
                "composer": composer,
                "copyrightMessage": copyright_message,
                "label": label,
                "artistPicture": artist_picture,
                "spotifyLink": spotify_link,
                "lastfmLink": lastfm_link,
                "tidalLink": tidal_link,
                "appleMusicLink": apple_music_link,
                "youtubeLink": youtube_link,
                "deezerLink": deezer_link,
                "timestamp": timestamp,
            }
        )
        result = await self._transport.procedure(
            "app.rocksky.scrobble.createScrobble", body=body, auth=True
        )
        return result or {}


def _drop_none(payload: dict[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in payload.items() if v is not None}
